### PR TITLE
[InstCombine] Reduce nested logical operator if poison is implied

### LINKED ISSUE #76623: [InstCombine] Missed optimization: fold `other_cond && (a s> -1) && a s< b` to `other_cond && a u< b` when `b` is a specific constant
Alive2 proof: https://alive2.llvm.org/ce/z/DT9eDZ

### Motivating example:

```llvm
define i1 @src(i32 %a, i1 %other_cond) {
entry:                                      
  %cmp1 = icmp sgt i32 %a, -1
  %or.cond = select i1 %other_cond, i1 %cmp1, i1 false
  %cmp2 = icmp slt i32 %a, 10086
  %ret = select i1 %or.cond, i1 %cmp2, i1 false
  ret i1 %ret
}
```
can be folded to:
```llvm
define i1 @tgt(i32 %a, i1 %other_cond) {
entry:                                      
  %or.cond = icmp ult i32 %a, 10086
  %ret = select i1 %other_cond, i1 %or.cond, i1 false
  ret i1 %ret
}
```
However, it folds if we replace the logical `and` with a bitwise `and`. I thought that Reassociate should be to blame, but [godbolt](https://godbolt.org/z/fY5vPx951) told me that similar fold for bitwise `and` is handled by InstCombine. Therefore, I tag InstCombine here.

### Real-world motivation

This snippet of IR is derived from [qemu/net/util.c@net_parse_macaddr](https://github.com/qemu/qemu/blob/7425b6277f12e82952cede1f531bfc689bf77fb1/net/util.c) (after O3 pipeline).
The example above is a reduced version. If you're interested in the original suboptimal IR and optimal IR, see also:https://godbolt.org/z/Ea8GGxKba

**Let me know if you can confirm that it's an optimization opportunity, thanks.**

### PR DESCRIPTION
**Pull request by XChy**
Fixes #76623
Alive2 proof: https://alive2.llvm.org/ce/z/gX6znJ  (I'm not sure how to write a proof for such transform, maybe there are mistakes)

In most cases, `icmp(a, C1) && (other_cond && icmp(a, C2))` will be reduced to `icmp(a, C1) & (other_cond && icmp(a, C2))`, since latter icmp always implies the poison of the former. After reduction, it's easier to simplify the icmp chain.
Similarly, this patch does the same thing for `(A && B) && C --> A && (B & C)`. Maybe we could constraint such reduction only on icmps if there is regression in benchmarks.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 1]
```diff
@@ -2985,6 +2985,14 @@ Instruction *InstCombinerImpl::foldSelectOfBools(SelectInst &SI) {
       return BinaryOperator::CreateOr(CondVal, FalseVal);
     }
 
+    if (match(CondVal, m_LogicalOr(m_Value(A), m_Value(B))) &&
+        isa<SelectInst>(CondVal) && impliesPoison(FalseVal, B)) {
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2989]:
```suggestion
    if (match(CondVal, m_Select(m_Value(A), m_One(), m_Value(B))) && impliesPoison(FalseVal, B)) {
```
Please add some negative tests:
+ multi-use
+ bitwise or

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 2]
```diff
@@ -2985,6 +2985,14 @@ Instruction *InstCombinerImpl::foldSelectOfBools(SelectInst &SI) {
...
+      // (A || B) || C --> A || (B | C)
+      return replaceInstUsesWith(
+          SI, Builder.CreateLogicalOp(Instruction::Or, A,
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2992]:
```suggestion
          SI, Builder.CreateLogicalOr(A,
```

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 3]
```diff
@@ -2985,6 +2985,13 @@ Instruction *InstCombinerImpl::foldSelectOfBools(SelectInst &SI) {
       return BinaryOperator::CreateOr(CondVal, FalseVal);
     }
 
+    if (match(CondVal, m_OneUse(m_Select(m_Value(A), m_One(), m_Value(B)))) &&
+        impliesPoison(FalseVal, B)) {
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2989]: Can you check the compile time impact?

Comment (XChy) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2989]: @nikic could you test the compile time impact of this patch?

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2989]: Sorry for the delay, here are the results: https://llvm-compile-time-tracker.com/compare.php?from=03ffb82c9e0d363c97ca37ede46719236616c88e&to=79752c9ee833172600523f038a6f83d64f657f9d&stat=instructions:u

**Review by nikic — APPROVED** [review]
LGTM. There is some small compile-time impact, but I think this is an important fold to have.

**Pull request merged**